# 🎬Getting familiar with data

In [2]:
import pandas as pd
import os
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [3]:
print_filler = '=================================\n'

## 🔹Analyse and transform dataset 

#### 💡Load full dataset, describe dataset, feature selection based on relevance, transform columns, explore chosen data with visualizations

In [4]:
file_name = 'Eastern flyway spring migration of adult white storks (data from Rotics et al. 2018).csv'
print(os.path.join('data',file_name))
data = pd.read_csv(os.path.join('data',file_name))
print(data.shape)
print(data.columns)

data\Eastern flyway spring migration of adult white storks (data from Rotics et al. 2018).csv
(2467077, 21)
Index(['event-id', 'visible', 'timestamp', 'location-long', 'location-lat',
       'comments', 'eobs:acceleration-axes',
       'eobs:acceleration-sampling-frequency-per-axis',
       'eobs:accelerations-raw', 'eobs:status', 'eobs:type-of-fix',
       'ground-speed', 'heading', 'height-above-ellipsoid', 'start-timestamp',
       'study-specific-measurement', 'sensor-type',
       'individual-taxon-canonical-name', 'tag-local-identifier',
       'individual-local-identifier', 'study-name'],
      dtype='object')


In [5]:
display(data.head())

event-id  visible                timestamp  location-long  location-lat  \
0  7335935935     True  2012-02-14 02:00:44.000      36.701858      3.985513   
1  7335935936     True  2012-02-14 02:05:10.000      36.701850      3.985472   
2  7335935937     True  2012-02-14 02:10:12.000      36.701815      3.985495   
3  7335935938     True  2012-02-14 02:15:11.000      36.701855      3.985515   
4  7335935939     True  2012-02-14 02:20:10.000      36.701822      3.985484   

   comments eobs:acceleration-axes  \
0       NaN                    NaN   
1       NaN                    NaN   
2       NaN                    NaN   
3       NaN                    NaN   
4       NaN                    NaN   

   eobs:acceleration-sampling-frequency-per-axis eobs:accelerations-raw  \
0                                            NaN                    NaN   
1                                            NaN                    NaN   
2                                            NaN                    NaN   
3                                            NaN                    NaN   
4                                            NaN                    NaN   

  eobs:status  ...  ground-speed  heading  height-above-ellipsoid  \
0           A  ...          0.02   358.48                   670.0   
1           A  ...          0.06   349.31                   667.9   
2           A  ...          0.08     0.00                   668.8   
3           A  ...          0.17     9.83                   666.7   
4           A  ...          0.11   336.86                   667.6   

   start-timestamp study-specific-measurement  sensor-type  \
0              NaN                        NaN          gps   
1              NaN                        NaN          gps   
2              NaN                        NaN          gps   
3              NaN                        NaN          gps   
4              NaN                        NaN          gps   

  individual-taxon-canonical-name tag-local-identifier  \
0                 Ciconia ciconia                 1787   
1                 Ciconia ciconia                 1787   
2                 Ciconia ciconia                 1787   
3                 Ciconia ciconia                 1787   
4                 Ciconia ciconia                 1787   

   individual-local-identifier  \
0                   1787/HH582   
1                   1787/HH582   
2                   1787/HH582   
3                   1787/HH582   
4                   1787/HH582   

                                          study-name  
0  Eastern flyway spring migration of adult white...  
1  Eastern flyway spring migration of adult white...  
2  Eastern flyway spring migration of adult white...  
3  Eastern flyway spring migration of adult white...  
4  Eastern flyway spring migration of adult white...  

[5 rows x 21 columns]

In [6]:
display(data.describe())

event-id  location-long  location-lat      comments  \
count  2.467077e+06   1.207130e+06  1.207130e+06  1.259947e+06   
mean   7.251066e+09   2.706471e+01  2.720805e+01  1.362054e+00   
std    8.366375e+07   7.560671e+00  1.831143e+01  1.806803e+00   
min    7.168545e+09   1.041956e+01 -3.058608e+01  1.276000e-01   
25%    7.169162e+09   2.289510e+01  1.263261e+01  2.808095e-01   
50%    7.169779e+09   2.897990e+01  2.706366e+01  7.644740e-01   
75%    7.336526e+09   3.301696e+01  4.330978e+01  1.810350e+00   
max    7.337143e+09   4.012790e+01  5.411720e+01  2.742710e+01   

       eobs:acceleration-sampling-frequency-per-axis  eobs:type-of-fix  \
count                                   1.259947e+06         1207130.0   
mean                                    1.054000e+01               3.0   
std                                     3.076030e-10               0.0   
min                                     1.054000e+01               3.0   
25%                                     1.054000e+01               3.0   
50%                                     1.054000e+01               3.0   
75%                                     1.054000e+01               3.0   
max                                     1.054000e+01               3.0   

       ground-speed       heading  height-above-ellipsoid  \
count  1.207130e+06  1.207130e+06            1.207130e+06   
mean   3.863259e+00  1.811902e+02            6.158689e+02   
std    5.991217e+00  1.522742e+02            5.602873e+02   
min    0.000000e+00  0.000000e+00           -2.564800e+03   
25%    1.300000e-01  1.507000e+01            2.024000e+02   
50%    3.000000e-01  1.861200e+02            4.421000e+02   
75%    7.980000e+00  3.447200e+02            9.017000e+02   
max    3.616200e+02  3.597900e+02            5.978000e+04   

       study-specific-measurement  tag-local-identifier  
count                2.345373e+06          2.467077e+06  
mean                 3.653552e+07          2.300970e+03  
std                  2.012813e+07          2.536245e+02  
min                  3.546255e+06          1.787000e+03  
25%                  1.064474e+07          2.306000e+03  
50%                  3.820403e+07          2.322000e+03  
75%                  5.527232e+07          2.341000e+03  
max                  6.238497e+07          2.859000e+03

In [7]:
null_counts = data.isnull().sum()
display(null_counts)
rows_without_null = data.dropna().shape[0]
print(f"{print_filler}Number of rows without null values:", rows_without_null)

event-id                                               0
visible                                                0
timestamp                                              0
location-long                                    1259947
location-lat                                     1259947
comments                                         1207130
eobs:acceleration-axes                           1207130
eobs:acceleration-sampling-frequency-per-axis    1207130
eobs:accelerations-raw                           1207130
eobs:status                                      1259947
eobs:type-of-fix                                 1259947
ground-speed                                     1259947
heading                                          1259947
height-above-ellipsoid                           1259947
start-timestamp                                  1207130
study-specific-measurement                        121704
sensor-type                                            0
individual-taxon-canonical-name

Number of rows without null values: 0


In [8]:
processed_data = data[['tag-local-identifier','timestamp','location-long','location-lat']]
null_counts = processed_data.isnull().sum()
display(null_counts)
rows_without_null = processed_data.dropna().shape[0]
print(f"{print_filler}Number of rows without null values:", rows_without_null)

tag-local-identifier          0
timestamp                     0
location-long           1259947
location-lat            1259947
dtype: int64

Number of rows without null values: 1207130


⚠️ The relevant columns we decide to keep are: `['tag-location-identifier','timestamp','location-long','location-lat']`

⚠️ We observe 1259947 records without latitude and longitude information, since the missing values cannot be approximated and the dataset size is decent in size after dropping those records, we proceed to drop them 

In [41]:
processed_data = processed_data.dropna()
print(processed_data.shape)

(1207130, 4)


In [42]:
processed_data.dtypes

tag-local-identifier      int64
timestamp                object
location-long           float64
location-lat            float64
dtype: object

In [44]:
print(list(processed_data['timestamp'])[0])

2012-02-14 02:00:44.000


In [45]:
processed_data['timestamp']=pd.to_datetime(processed_data['timestamp'])
processed_data.dtypes

tag-local-identifier             int64
timestamp               datetime64[ns]
location-long                  float64
location-lat                   float64
dtype: object

In [55]:
processed_data.to_csv(os.path.join('data','StorkMigration.csv'))

In [58]:
selected_id = list(processed_data['tag-local-identifier'])[0]
selected_data_id = processed_data[processed_data['tag-local-identifier']==selected_id]
print(selected_data_id.shape)
print(selected_data_id['timestamp'])

(19172, 4)
0       2012-02-14 02:00:44
1       2012-02-14 02:05:10
2       2012-02-14 02:10:12
3       2012-02-14 02:15:11
4       2012-02-14 02:20:10
                ...        
19167   2013-04-21 19:40:16
19168   2013-04-21 20:00:13
19169   2013-04-21 20:20:31
19170   2013-04-21 20:40:32
19171   2013-04-21 21:00:12
Name: timestamp, Length: 19172, dtype: datetime64[ns]


In [76]:
for stork in set(processed_data['tag-local-identifier']):
    selected_data_id = processed_data[processed_data['tag-local-identifier']==stork]
    result = selected_data_id.groupby([selected_data_id['timestamp'].dt.year]).count()
    
    display(result)

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      17498      17498          17498         17498
2013                      14010      14010          14010         14010
2014                       7917       7917           7917          7917
2015                       4613       4613           4613          4613

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      11988      11988          11988         11988
2013                       9733       9733           9733          9733
2014                      12437      12437          12437         12437

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      10042      10042          10042         10042
2014                      10527      10527          10527         10527

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      16172      16172          16172         16172
2013                      11179      11179          11179         11179
2014                       9956       9956           9956          9956

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      11826      11826          11826         11826
2014                      10669      10669          10669         10669
2015                      11803      11803          11803         11803

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      13635      13635          13635         13635
2014                      12165      12165          12165         12165
2015                      13824      13824          13824         13824
2016                      12974      12974          12974         12974

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      11294      11294          11294         11294
2014                      10781      10781          10781         10781
2015                      16328      16328          16328         16328
2016                       9578       9578           9578          9578

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      13537      13537          13537         13537
2014                      10924      10924          10924         10924
2015                      14553      14553          14553         14553

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      12478      12478          12478         12478

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      16372      16372          16372         16372
2014                      13449      13449          13449         13449
2015                      18790      18790          18790         18790
2016                      13023      13023          13023         13023

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      12508      12508          12508         12508
2014                      13290      13290          13290         13290
2016                      11939      11939          11939         11939

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      16066      16066          16066         16066
2014                      22321      22321          22321         22321
2015                      13454      13454          13454         13454
2016                      15892      15892          15892         15892

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      16761      16761          16761         16761

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      15643      15643          15643         15643
2014                      14265      14265          14265         14265

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      15465      15465          15465         15465

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      11701      11701          11701         11701
2015                      14598      14598          14598         14598

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      11704      11704          11704         11704
2014                      16962      16962          16962         16962
2015                      20267      20267          20267         20267
2016                      16265      16265          16265         16265

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      14148      14148          14148         14148

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      12475      12475          12475         12475
2014                      13034      13034          13034         13034
2015                      13002      13002          13002         13002

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                       9464       9464           9464          9464

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      12903      12903          12903         12903
2014                      12976      12976          12976         12976
2015                      13892      13892          13892         13892
2016                      12807      12807          12807         12807

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      13248      13248          13248         13248
2014                      13459      13459          13459         13459

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      11696      11696          11696         11696
2014                      11967      11967          11967         11967
2015                      13301      13301          13301         13301

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      14966      14966          14966         14966
2014                      10791      10791          10791         10791
2015                      15181      15181          15181         15181
2016                      16974      16974          16974         16974

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      13569      13569          13569         13569

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      11283      11283          11283         11283
2015                      10233      10233          10233         10233

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      14909      14909          14909         14909
2015                      13760      13760          13760         13760
2016                      14726      14726          14726         14726

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      15874      15874          15874         15874
2013                      15386      15386          15386         15386
2014                      11718      11718          11718         11718
2015                      13254      13254          13254         13254

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                       9660       9660           9660          9660

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      11301      11301          11301         11301

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      13947      13947          13947         13947
2015                      12487      12487          12487         12487
2016                       9210       9210           9210          9210

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2014                      19003      19003          19003         19003
2015                      13370      13370          13370         13370
2016                      11931      11931          11931         11931

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2012                      10410      10410          10410         10410
2013                       8762       8762           8762          8762

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      15285      15285          15285         15285
2014                      12891      12891          12891         12891
2015                      15432      15432          15432         15432
2016                      16674      16674          16674         16674

tag-local-identifier  timestamp  location-long  location-lat
timestamp                                                              
2013                      12595      12595          12595         12595

In [70]:
selected_id = set(processed_data['tag-local-identifier'])
print(selected_id)

{1793, 1794, 2305, 1796, 1797, 2306, 2307, 2313, 2317, 2319, 2320, 2322, 2323, 2327, 2840, 2841, 2330, 2331, 2332, 2842, 2334, 2336, 2337, 2341, 2342, 2346, 2859, 1791, 2491, 2492, 2499, 2421, 1787, 2301, 2303}
